### DISCUSSION TITLE
Default `kamal rollback` to prior deploy

### DISCUSSION THREAD

**Discussion by motdotla** [basecamp/kamal#621] [Ideas]
The typical rollback flow is:

1. Oh shoot, my deploy succeeded but my container is down
2. Run `kamal app containers`
3. Identify 2nd from top container commit-hash
4. Run `kamal rollback commit-hash`
5. Yay, my app is back up

It would be nice if `kamal rollback` (without any arguments) did the above automatically. This is what heroku rollback does.

**Comment by dhh** [comment]
I like that. Please do investigate.

**Reply by DanielJackson-Oslo** [reply]
@dhh I just ran into this, and would love to contribute.

So far, it seems like there are three issues with kamal rollback, that could all be fixed by reworking how it identifies which container to rollback to:

1) It rolls back to a version that it finds based looking up the container id using the container hash, instead of just being passed the container id directly, even though the container id is easier to find to find with `kamal  app containers`. 

This also causes issues when there are no code changes, only configuration changes (changed .env file, for example) – where, if I understand correctly, the _container hash is the same_, even though the container has a different id and different contents. 

Here is a real example from my app:

```
CONTAINER ID   IMAGE                                                                      COMMAND                  CREATED             STATUS                           PORTS     NAMES
ee3ce6366895   danieljacksonnor/lokaler-lnu-no:724671c68992aff92349f7599afe2274eaed9a10   "/rails/bin/docker-e…"   9 minutes ago       Up 9 minutes                     80/tcp    staging-web-staging-724671c68992aff92349f7599afe2274eaed9a10
8b71ff79378b   danieljacksonnor/lokaler-lnu-no:724671c68992aff92349f7599afe2274eaed9a10   "/rails/bin/docker-e…"   15 minutes ago      Exited (255) 9 minutes ago                 staging-web-staging-724671c68992aff92349f7599afe2274eaed9a10_replaced_a0c676aea58949b2
2b97f624d657   danieljacksonnor/lokaler-lnu-no:724671c68992aff92349f7599afe2274eaed9a10   "/rails/bin/docker-e…"   28 minutes ago      Exited (255) 15 minutes ago                staging-web-staging-724671c68992aff92349f7599afe2274eaed9a10_replaced_bb729b5814973242
faed64ff269f   danieljacksonnor/lokaler-lnu-no:724671c68992aff92349f7599afe2274eaed9a10   "/rails/bin/docker-e…"   About an hour ago   Exited (255) 28 minutes ago                staging-web-staging-724671c68992aff92349f7599afe2274eaed9a10_replaced_aa4832b05ff49565
```

The `8b71ff79378b` deploy was dead, due to a misconfigured .env file. I wanted to rollback to container id `2b97f624d657`, but had to use the hash `724671c68992aff92349f7599afe2274eaed9a10` which it shared with the all of the last deploys and thus the rollback just gave me the same version I already had. 

2) You have to specify a version at all, instead of just panic hitting `kamal rollback` to undo the latest fuck-up.

3) There is no `kamal rollback --help` (in fact it tries to rollback to a container with the hash "--help" if you try)

-

My suggestion:

a) If you do `kamal rollback SOME_ID_OR_HASH`, it looks up to see if it can find a matching container_id _or_ container hash to rollback to. (Alternatively you could be explicit about what data you are passing: `kamal rollback id=SOME_ID` or `kamal rollback hash=SOME_HASH`)

b) If you just do `kamal rollback` it gives you a list of the potential rollback candidates and asks you to pick one to rollback to, defaulting to the previous one. (I think I prefer this over just rolling back to the last version mindlessly, even though it's more complex to code up I think it gives more support and clarity in what is always a stressful situation: Trying to rollback a fuck up)

c) And add a simple --help explaining this.

Does that plan sound sane? Should I explore making it real?

**Comment by dhh** [comment]
Hey Daniel,

Thanks for wanting to work on this. I'm most excited about b, since then you don't first have to look at an inventory elsewhere to do the rollback. Here's what it could show based on your example if you just do `kamal rollback`:

```
Rollback to deployment from:

1) 15 minutes ago (8b71ff79378b)
2) 28 minutes ago (2b97f624d657)
3) About an hour ago (faed64ff269f)

Pick: [1]
```

And you should be able to also do `kamal rollback --container-id=2b97f624d657`.
